tests/subsys/emds/emds_flash/src/main.c

```diff
@@ -21,661 +22,502 @@
 #endif
 
-#define ADDR_OFFS_MASK 0x0000FFFF
-
-static struct {
-	const struct device *fd;
-	uint32_t offset;
-	uint32_t size;
-	uint32_t ate_idx_start;
-	uint32_t data_wra_offset;
-} m_test_fd;
-
-/* Allocation Table Entry */
-struct test_ate {
-	uint16_t id;       /* data id */
-	uint16_t offset;   /* data offset within sector */
-	uint16_t len;      /* data len within sector */
-	uint8_t crc8_data; /* crc8 check of the entry */
-	uint8_t crc8;      /* crc8 check of the entry */
-} __packed;
-
-static struct emds_fs ctx;
-static uint16_t m_sec_size;
-static uint16_t m_sec_cnt;
-static const struct flash_area *m_fa;
+#define EMDS_SNAPSHOT_METADATA_MARKER 0x4D444553
+#define PARTITIONS_NUM_MAX 2
 
-/** Local functions ***********************************************************/
+static struct emds_partition partition[PARTITIONS_NUM_MAX];
 
-static int flash_block_cmp(uint32_t addr, const void *data, size_t len)
+/** Local functions ***********************************************************/
+static void *emds_flash_setup(void)
 {
-	const uint8_t *data8 = (const uint8_t *)data;
-	int rc;
-	size_t bytes_to_cmp, block_size;
-	uint8_t buf[32];
-
-	block_size = sizeof(buf) & ~(4 - 1U);
-
-	while (len) {
-		bytes_to_cmp = MIN(block_size, len);
-		rc = flash_read(m_test_fd.fd, addr, buf, bytes_to_cmp);
-		if (rc) {
-			return rc;
-		}
-		rc = memcmp(data8, buf, bytes_to_cmp);
-		if (rc) {
-			return -ENOTSUP;
-		}
-
-		len -= bytes_to_cmp;
-		addr += bytes_to_cmp;
-		data8 += bytes_to_cmp;
+	const uint8_t id[] = {FIXED_PARTITION_ID(emds_partition_0),
+			      FIXED_PARTITION_ID(emds_partition_1)};
+
+	for (int i = 0; i < ARRAY_SIZE(id); i++) {
+		zassert_ok(flash_area_open(id[i], &partition[i].fa), "Failed to open flash area %d",
+			   id[i]);
+		zassert_ok(emds_flash_init(&partition[i]), "Failed to initialize flash area %d",
+			   id[i]);
 	}
 
-	return 0;
+	return NULL;
 }
 
-static int flash_cmp_const(uint32_t addr, uint8_t value, size_t len)
+static void emds_flash_partitions_erase(void *fixture)
 {
-	int rc;
-	size_t bytes_to_cmp, block_size;
-	uint8_t cmp[32];
-
-	block_size = sizeof(cmp) & ~(4 - 1U);
-
-	memset(cmp, value, block_size);
-	while (len) {
-		bytes_to_cmp = MIN(block_size, len);
-		rc = flash_block_cmp(addr, cmp, bytes_to_cmp);
+	(void)fixture;
 
-		if (rc) {
-			return rc;
-		}
-		len -= bytes_to_cmp;
-		addr += bytes_to_cmp;
+	for (int i = 0; i < PARTITIONS_NUM_MAX; i++) {
+		zassert_ok(emds_flash_erase_partition(&partition[i]),
+			   "Failed to erase partition %d", i);
 	}
-	return 0;
 }
 
-static inline size_t align_size(size_t len)
+static struct emds_snapshot_metadata *
+snapshot_make(const struct emds_partition *partition,
+	      struct emds_snapshot_metadata *previous_snapshot, off_t metadata_off,
+	      bool is_md_crc_correct, bool is_d_crc_correct, uint32_t fresh_cnt)
 {
-	return (len + (EMDS_FLASH_BLOCK_SIZE - 1U)) & ~(EMDS_FLASH_BLOCK_SIZE - 1U);
-}
+	static struct emds_snapshot_metadata metadata;
```

Comment:
Using static struct can produce some strange consequences since this is passed around almost everywhere in the test logic.

Reply:
That's not correct. In the first run "when function gets NULL" the structure us cleaned here: https://github.com/nrfconnect/sdk-nrf/pull/23380/files#diff-1afb6c02cc42eb9329c58abc857962b711bbdc08080606c431010c993f543a49R87-R89
in other runs, it is rewritten by data from incoming structure.
No strange consequences.